src/main/java/rife/engine/TomcatServer.java

```diff
@@ -0,0 +1,167 @@
+        roles_.put(role, user);
+        return this;
+    }
+
+    /**
+     * Add a user for the in-memory realm.
+     */
+    public TomcatServer addUser(String user, String pass) {
+        users_.put(user, pass);
+        return this;
+    }
+
+    /**
+     * Add the location of the webapp directory or WAR for the root context.
+     *
+     * @param docBase Base directory for the context, for static file.
+     */
+    public TomcatServer addWebapp(String docBase) {
+        docBase_ = docBase;
+        return this;
+    }
+
+    /**
+     * Configures the Tomcat base directory on which all others, such as the {@code work} directory, will be derived.
+     */
+    public TomcatServer baseDir(String dir) {
+        baseDir_ = dir;
+        return this;
+    }
+
+    /**
+     * Configures the host name the server will be listening on.
+     */
+    public TomcatServer hostname(String host) {
+        hostname_ = host;
+        return this;
+    }
+
+    /**
+     * Configures the port the server will be listening to.
+     */
+    public TomcatServer port(int port) {
+        port_ = port;
+        return this;
+    }
+
+    /**
+     * Retrieves the hierarchical properties for this server instance.
+     */
+    public HierarchicalProperties properties() {
+        return properties_;
+    }
+
+    /**
+     * Starts the embedded server.
+     */
+    public TomcatServer start(Site site) {
+        tomcat_ = new Tomcat();
+
+        if (baseDir_ == null) {
+            var tmpDir = new File(RifeConfig.global().getTempPath(), "rife2.tomcat." + port_);
+            tmpDir.deleteOnExit();
+            tomcat_.setBaseDir(tmpDir.getAbsolutePath());
+        } else {
+            tomcat_.setBaseDir(new File(baseDir_).getAbsolutePath());
+        }
+
+        var ctx = tomcat_.addContext("", new File(docBase_).getAbsolutePath());
+
+        var servletName = "default-servlet";
+        var defaultServlet = new DefaultServlet();
+        Tomcat.addServlet(ctx, servletName, defaultServlet);
+        ctx.addServletMappingDecoded("/*", servletName);
+
+        var filterName = "RIFE2";
+        var rifeFilter = new RifeFilter();
+        rifeFilter.init(properties_, site);
+
+        var filerDef = new FilterDef();
+        filerDef.setFilter(rifeFilter);
+        filerDef.setFilterName(filterName);
+        ctx.addFilterDef(filerDef);
+
+        var filterMap = new FilterMap();
+        filterMap.setFilterName(filterName);
+        filterMap.addURLPattern("/*");
+        ctx.addFilterMap(filterMap);
+
+        tomcat_.setPort(port_);
+
+        if (hostname_ != null) {
+            tomcat_.setHostname(hostname_);
+        }
+
+        users_.forEach((u, p) -> tomcat_.addUser(u, p));
+
+        roles_.forEach((u, r) -> tomcat_.addRole(u, r));
+
+        // Tomcat opens the port only if called at least once
+        tomcat_.getConnector();
+
+        try {
+            tomcat_.start();
+//            tomcat_.getServer().await();
```

Comment:
Should this comment still be here?

Reply:
Yeah, it's a reminder in case I need to change things around.

Reply:
Could you add a comment that explains why this is a reminder and is good to stay commented?

Reply:
Removed. After more tinkering, I don't think we'll ever need to use that.